modules/table/src/integrationTest/java/org/apache/ignite/internal/table/ItTxResourcesVacuumTest.java

```diff
@@ -0,0 +1,798 @@
+
+                cleanupAllowedFut.join();
+
+                return true;
+            }
+
+            return false;
+        });
+
+        CompletableFuture<Void> commitFut = tx.commitAsync();
+
+        assertThat(cleanupStarted, willCompleteSuccessfully());
+
+        // Wait for volatile tx state vacuum. This is possible because tx finish is complete.
+        waitForTxStateVacuum(txId, commitPartId, false, 10_000);
+
+        log.info("Test: volatile state vacuumized");
+
+        cleanupAllowedFut.complete(null);
+
+        cleanupAllowed[0] = true;
+
+        assertThat(commitFut, willCompleteSuccessfully());
+
+        waitForTxStateVacuum(txId, commitPartId, true, 10_000);
+
+        // Trying to read the data.
+        Tuple data = view.get(null, Tuple.create().set("key", tuple.longValue("key")));
+        assertEquals(tuple, data);
+    }
+
+    /**
+     * Checks that the tx recovery doesn't change tx finish result from COMMITTED to ABORTED if it once saved in the persistent storage.
+     *
+     * <ul>
+     *     <li>Start a transaction tx0;</li>
+     *     <li>Upsert some value;</li>
+     *     <li>Block {@link TxCleanupMessage}-s;</li>
+     *     <li>Start the commit of tx0 and with for tx state COMMITTED to be replicated in persistent storage;</li>
+     *     <li>Stop the tx0's coordinator;</li>
+     *     <li>Wait for tx0's state vacuum;</li>
+     *     <li>Try to get the data that has been committed by tx0, ensure the data is correct.</li>
+     * </ul>
+     */
+    @Test
+    public void testRecoveryAfterPersistentStateVacuumized() throws InterruptedException {
+        // This node isn't going to be stopped, so let it be node 0.
+        IgniteImpl commitPartitionLeaseholder = cluster.node(0);
+
+        Tuple tuple0 = findTupleToBeHostedOnNode(commitPartitionLeaseholder, TABLE_NAME, null, INITIAL_TUPLE, NEXT_TUPLE, true);
+
+        int commitPartId = partitionIdForTuple(commitPartitionLeaseholder, TABLE_NAME, tuple0, null);
+
+        Set<String> commitPartitionNodes = partitionAssignment(commitPartitionLeaseholder,
+                new TablePartitionId(tableId(commitPartitionLeaseholder, TABLE_NAME), commitPartId));
+
+        // Choose some node that doesn't host the partition as a tx coordinator.
+        IgniteImpl coord0 = findNode(n -> !commitPartitionNodes.contains(n.name()));
+
+        RecordView<Tuple> view0 = coord0.tables().table(TABLE_NAME).recordView();
+
+        // Put some value into the table.
+        Transaction tx0 = coord0.transactions().begin();
+        UUID txId0 = txId(tx0);
+
+        log.info("Test: Transaction 0 [tx={}].", txId0);
+
+        log.info("Test: Commit partition of transaction 0 [leaseholder={}, hostingNodes={}].", commitPartitionLeaseholder.name(),
+                commitPartitionNodes);
+
+        view0.upsert(tx0, tuple0);
+
+        CompletableFuture<Void> cleanupStarted = new CompletableFuture<>();
+        boolean[] cleanupAllowed = new boolean[1];
```

Comment:
Contents of array 'cleanupAllowed' are read, but never written to

Reply:
removed